### ISSUE TITLE
Uri Authority instances with empty host but non-empty port or user-info not handled correctly

### ISSUE THREAD

**Issue by sirthias** [spray/spray#1014]
Backport of https://github.com/akka/akka/issues/16953.
Fix in https://github.com/akka/akka/pull/16954.

Easy to backport.

**Comment by wjur** [comment]
Could you please tell me if anyone is working on it? I would like to contribute and backport it by myself but I am not sure if the ticked is already "taken".

**Comment by sirthias** [comment]
No one from our side is currently working on this.
So it'd be great you picked this up...
Thanks in advance!

**Comment by wjur** [comment]
I've started from analysing the fixing commit in akka and one thing confused me. 
Code:
````Scala
"normal examples" in {
    resolve("g:h") shouldEqual "g:h"
````
changes to: 
````Scala
"normal examples" in {
    resolve("g:h") shouldEqual "g://h"
````
While the RFC says (in "normal examples")
> "g:h"           =  "g:h"

For me, it seems that the commit breaks compliance with the RFC. Maybe I just don't get it. Could you explain it to me a bit? 
I backported the fix to current master and still one test fails. Now, I'm wondering whether I should change resolve("g:h") === "g:h" to resolve("g:h") === "g://h" in the other part of the tests, too (As this kind of tests do not exist in akka).

**Comment by sirthias** [comment]
We had a discussion in https://github.com/akka/akka/pull/16954 about this question (click on `Show outdated diff`). Essentially it comes down to this sentence from [this section of RFC3986](http://tools.ietf.org/html/rfc3986#section-6.2.3):

> Normalization should not remove delimiters when their associated
   component is empty unless licensed to do so by the scheme specification.

If the case you highlighted the URI has a scheme (`g`), no authority and a "rootless" path (`path-rootless`). The patch in https://github.com/akka/akka/pull/16954 always renders the `//` delimiter for empty authorities, if the scheme is not empty or `mailto`. As your example shows, this is not quite good enough. In your example we are not allowed to render the `//` delimiter because this would turn what is a rootless path into a host! So we need to apply a small fix to https://github.com/akka/akka/pull/16954, which prevents the rendering of the `//` delimiter if the path is rootless.

So, to recap, we need to render the `//` between a non-empty scheme and an empty authority exactly when these conditions are met:

1. the scheme is not `mailto` (potentially more exception schemes will be added here later)
2. the path is non-empty and starts with a segment (not a slash)

So, thanks for bringing this up as it shows a flaw in https://github.com/akka/akka/pull/16954 that we need to fix. After getting this right in spray, would you also want to put it an Akka PR with the correction (and maybe some more tests that we are apparently missing in Akka)?

**Comment by wjur** [comment]
I have prepared my RP. Please take a look the comment in the code as I am not sure if we understand the flaw in the same way.
I will appreciate your comments, too (I am willing to polish it so do not hesitate to point out my mistakes or any other doubts).

Yes, I would like to contribute to Akka, too. I would be proud :)
